### PR TITLE
Add runtime_context to get some runtime fields in worker

### PR DESCRIPTION
**Pull request by jovany-wang**
## What do these changes do?
Usage:
```python
# in worker or driver
id = ray.get_runtime_context().current_driver_id
```

We add a `runtime_cotext` to keep some runtime fields.
In this PR, we add an attribution `current_driver_id` which is used in many situations as we needed.

In the future, we can add other attrs to this like `current_actor_id`, `was_current_actor_reconstructed` ant etc.

## Related issue number
N/A

### CODE REVIEW

File: test/runtest.py [hunk 1]
```diff
@@ -2767,3 +2767,17 @@ def f():
...
+    ray.init(driver_id=specified_driver_id)
+
+    # in driver
+    assert_equal(specified_driver_id, ray.runtime_context.current_driver_id)
+
+    # in worker
+    @ray.remote
+    def f():
+        return ray.runtime_context.current_driver_id.id()
+    assert_equal(specified_driver_id, ray.ObjectID(ray.get(f.remote())))
```
Comment (robertnishihara) [inline test/runtest.py]:
```suggestion
    assert specified_driver_id == ray.ObjectID(ray.get(f.remote()))
```

File: test/runtest.py [hunk 2]
```diff
@@ -2767,3 +2767,17 @@ def f():
...
 
     assert ray.get(f.remote()) == 1
+
+
+def test_runtime_context(shutdown_only):
+    specified_driver_id = ray.ObjectID(b"00112233445566778899")
+    ray.init(driver_id=specified_driver_id)
+
+    # in driver
+    assert_equal(specified_driver_id, ray.runtime_context.current_driver_id)
```
Comment (robertnishihara) [inline test/runtest.py]:
```suggestion
    assert specified_driver_id == ray.runtime_context.current_driver_id
```

File: test/runtest.py [hunk 3]
```diff
@@ -2767,3 +2767,17 @@ def f():
         return 1
 
     assert ray.get(f.remote()) == 1
+
+
+def test_runtime_context(shutdown_only):
+    specified_driver_id = ray.ObjectID(b"00112233445566778899")
+    ray.init(driver_id=specified_driver_id)
```
Comment (robertnishihara) [inline test/runtest.py]:
```suggestion
    ray.init(num_cpus=1, driver_id=specified_driver_id)
```

File: python/ray/runtime_context.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+
+class RuntimeContext(object):
+    """A class used for getting runtime context.
+    """
```
Comment (robertnishihara) [inline python/ray/runtime_context.py]: If the docstring is a single line, the `"""` should be on the previous line.

File: python/ray/runtime_context.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    def __init__(self, worker=None):
+        self.worker = worker
+
+    @property
+    def current_driver_id(self):
+        """Get current driver id for this worker or driver.
```
Comment (robertnishihara) [inline python/ray/runtime_context.py]: Change the wording, and formatting to

```python
        """Get current driver ID for this worker or driver.

        Returns:
            If called by a driver, this returns the driver ID. If called in
                a task, return the driver ID of the associated driver.
        """
```

**Review by robertnishihara — COMMENTED** [review]
What is the use case that you have in mind for this? Are you planning on adding more context beyond the driver ID?

There are some alternatives, e.g., discussed in https://github.com/ray-project/ray/pull/3805 of simply exposing the current task, and then the driver ID could be accessed through the task.

**Comment by raulchen** [comment]
@robertnishihara The purpose of this PR is to expose standard APIs for some useful runtime info, e.g., current task, driver id, redis address, etc. 

Previously, users need to get these data from worker's internal fields or methods, which is likely to break as the code changes. So I'd like to propose exposing some standard APIs for them. E.g.,
```
ray.runtime_context.current_task()
ray.runtime_context.current_driver_id()
ray.runtime_context.redis_address()
```

What do you think?

File: python/ray/__init__.py [hunk 6]
```diff
@@ -122,6 +123,7 @@
     "remote",
     "shutdown",
     "wait",
+    "runtime_context",
```
Comment (raulchen) [inline python/ray/__init__.py]: Could you keep these items in alphabetical order?

**Comment by jovany-wang** [comment]
I have refined this PR, feel free to review it agagin.

File: test/runtest.py [hunk 7]
```diff
@@ -2842,3 +2842,18 @@ def f():
         return 1
 
     assert ray.get(f.remote()) == 1
+
+
+def test_runtime_context(shutdown_only):
```
Comment (raulchen) [inline test/runtest.py]: we can change `runtest.py/test_specific_driver_id` to use runtime context, then we don't need to add this test.

**Review by robertnishihara — COMMENTED** [review]
I think this is a fine approach to take. I prefer `ray._get_runtime_context()` since presumably this will change a bunch in the future (e.g., we will add new fields, change the name of some fields, etc).

Almost all users shouldn't need to think about the runtime context, right? Certainly the current driver ID doesn't matter for most users. So I think it makes sense for this to be somewhat less exposed.

**Comment by raulchen** [comment]
Sounds good. Agreed that this should be less exposed.

**Review by raulchen — APPROVED** [review]

**Pull request merged**